File: app/src/main/java/com/example/test/mvvmsampleapp/service/repository/ProjectRepository.java
package com.example.test.mvvmsampleapp.service.repository;

import com.example.test.mvvmsampleapp.service.model.Project;
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

import java.util.List;

import io.reactivex.Single;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ProjectRepository {
    private GitHubService gitHubService;
    private static ProjectRepository projectRepository;

    private ProjectRepository() {
        //TODO this gitHubService instance will be injected using Dagger in part #2 ...
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(GitHubService.HTTPS_API_GITHUB_URL)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        gitHubService = retrofit.create(GitHubService.class);
    }

    public synchronized static ProjectRepository getInstance() {
        //TODO No need to implement this singleton in Part #2 since Dagger will handle it ...
        if (projectRepository == null) {
            if (projectRepository == null) {
                projectRepository = new ProjectRepository();
            }
        }
        return projectRepository;
    }

    public Single<List<Project>> getProjectList(String userId) {
        return gitHubService.getProjectList(userId);
    }

    public Single<Project> getProjectDetails(String userID, String projectName) {
        simulateDelay();
        return gitHubService.getProjectDetails(userID, projectName);
    }

    private void simulateDelay() {
        try {
            Thread.sleep(10);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/example/test/mvvmsampleapp/service/repository/ProjectRepository.kt
package com.example.test.mvvmsampleapp.service.repository

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import com.example.test.mvvmsampleapp.service.model.Project
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ProjectRepository @Inject
constructor(private val gitHubService: GitHubService) {

    fun getProjectList(userId: String): LiveData<List<Project>> {
        val data = MutableLiveData<List<Project>>()

        gitHubService.getProjectList(userId)
                .enqueue(object : Callback<List<Project>> {
                    override fun onResponse(call: Call<List<Project>>,
                                            response: Response<List<Project>>) {
                        data.value = response.body()
                    }

                    override fun onFailure(call: Call<List<Project>>,
                                           t: Throwable) {
                        // TODO better error handling in part #2 ...
                        data.value = null
                    }
                })

        return data
    }

    fun getProjectDetails(userID: String, projectName: String?): LiveData<Project> {
        val data = MutableLiveData<Project>()

        gitHubService.getProjectDetails(userID, projectName)
                .enqueue(object : Callback<Project> {
                    override fun onResponse(call: Call<Project>,
                                            response: Response<Project>) {
                        simulateDelay()
                        data.value = response.body()
                    }

                    override fun onFailure(call: Call<Project>,
                                           t: Throwable) {
                        // TODO better error handling in part #2 ...
                        data.value = null
                    }
                })

        return data
    }

    private fun simulateDelay() {
        try {
            Thread.sleep(10)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }
}
